BTF: arbitrary __attribute__ encoding

theihor opened the pull request.

theihor committed:
libbpf: introduce kflag for type_tags and decl_tags in BTF

Add the following functions to libbpf API:
* btf__add_type_attr()
* btf__add_decl_attr()

These functions allow to add to BTF the type tags and decl tags with
info->kflag set to 1. The kflag indicates that the tag directly
encodes an __attribute__ and not a normal tag.

See Documentation/bpf/btf.rst changes for details on the semantics.

Suggested-by: Andrii Nakryiko <andrii@kernel.org>
Signed-off-by: Ihor Solodrai <ihor.solodrai@pm.me>

theihor committed:
libbpf: check the kflag of type tags in btf_dump

If the kflag is set for a BTF type tag, then the tag represents an
arbitrary __attribute__. Change btf_dump accordingly.

Signed-off-by: Ihor Solodrai <ihor.solodrai@pm.me>

eddyz87 commented on tools/testing/selftests/bpf/prog_tests/btf.c:
```diff
@@ -4206,6 +4204,23 @@ static struct btf_raw_test raw_tests[] = {
 	.btf_load_err = true,
 	.err_str = "Type tags don't precede modifiers",
 },
+{
```
Note: usually for kernel patches new tests are added as separate commits, modifications to tests are included only to the extent to keep tests passing.

eddyz87 commented on Documentation/bpf/btf.rst:
```diff
@@ -499,12 +499,21 @@ the attribute is applied to a ``struct``/``union`` member or
+__attribute__. In this case, ``name_off`` encodes a string
+representing the attribute-list of the attribute specifier. For
+example, for an ``__attribute__((aligned(4)))`` the string is equal to
+``"aligned(4)"``.
```
Maybe emphasize that `"`s are not inserted.

eddyz87 commented on tools/lib/bpf/btf.c:
```diff
@@ -2113,12 +2113,17 @@ static int btf_add_ref_kind(struct btf *btf, int kind, const char *name, int ref
 	return btf_commit_type(btf, sz);
 }
 
+static inline int btf_add_ref_kind(struct btf *btf, int kind, const char *name, int ref_type_id)
```
nit: note sure about this function, would fix the callsites instead, tbh.

eddyz87 commented on tools/lib/bpf/btf_dump.c:
```diff
@@ -1494,7 +1494,10 @@ static void btf_dump_emit_type_chain(struct btf_dump *d,
 		case BTF_KIND_TYPE_TAG:
```
Nit: commit tag is usually `libbpf:`.

eddyz87 commented:
You modify dump to print attribute for type tags, but not for decl tags, while kernel side is adjusted to allow both. Maybe change dump for decl tag?

eddyz87 commented:
Also, maybe extend the `tools/testing/selftests/bpf/prog_tests/btf_dump.c` test case?

theihor committed:
bpf: allow kind_flag for BTF type and decl tags

BTF type tags and decl tags now may have info->kflag set to 1,
changing the semantics of the tag.

Change BTF verification to permit BTF that makes use of this feature:
  * remove kflag check in btf_decl_tag_check_meta(), as both values
    are valid
  * allow kflag to be set for BTF_KIND_TYPE_TAG type in
    btf_ref_type_check_meta()

Modify a selftest checking for kflag in decl_tag accordingly.

Signed-off-by: Ihor Solodrai <ihor.solodrai@pm.me>

theihor committed:
selftests/bpf: add a BTF verification test for kflagged type_tag

Add a BTF verification test case for a type_tag with a kflag set.
Type tags with a kflag are now valid.

Add BTF_DECL_ATTR_ENC and BTF_TYPE_ATTR_ENC test helper macros,
corresponding to *_TAG_ENC.

Signed-off-by: Ihor Solodrai <ihor.solodrai@pm.me>

theihor committed:
selftests/bpf: add a btf_dump test for type_tags

Factor out common routines handling custom BTF from
test_btf_dump_incremental. Then use them in the
test_btf_dump_type_tags.

test_btf_dump_type_tags verifies that a type tag is dumped correctly
with respect to its kflag.

Signed-off-by: Ihor Solodrai <ihor.solodrai@pm.me>
